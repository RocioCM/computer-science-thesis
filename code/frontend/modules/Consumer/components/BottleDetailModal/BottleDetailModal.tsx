import Modal, { ModalProps } from '@/common/components/Modal';
import ConsumerServices from '../../services';
import { toast } from 'react-toastify';
import { Material } from '../../types';
import { useEffect, useState } from 'react';
import LoadingSpinner from '@/common/components/LoadingSpinner';

export interface Props extends ModalProps {
  handleCancel: () => any;
  wasteBottleId: number | null;
}

const BottleDetailModal = ({
  handleCancel,
  wasteBottleId,
  ...props
}: Props) => {
  const [wasteBottle, setWasteBottle] = useState<any | null>(null);
  const [loading, setLoading] = useState(true);

  const fetchWasteBottle = async (wasteBottleId: number) => {
    setLoading(true);
    const { ok, data } = await ConsumerServices.getWasteBottleTracking(
      wasteBottleId
    );
    if (ok) {
      const stagesObject = data.reduce<Record<string, any>>((acc, curr) => {
        acc[curr.stage] = curr.data;
        return acc;
      }, {});
      const { ok, data: dataHistory } = await ConsumerServices.searchBottle(
        stagesObject.pickup?.trackingCode
      );
      if (ok) {
        dataHistory.forEach((item: any) => {
          stagesObject[item.stage] = item.data;
        });
        setWasteBottle(stagesObject);
      } else {
        toast.error(
          'Código no encontrado. Por favor, verifique e intente nuevamente'
        );
      }
    } else {
      toast.error(
        'Código no encontrado. Por favor, verifique e intente nuevamente'
      );
    }
    setLoading(false);
  };

  useEffect(() => {
    if (wasteBottleId) {
      fetchWasteBottle(wasteBottleId);
    }
  }, [wasteBottleId]);

  if (!wasteBottleId) {
    return null;
  }

  return (
    <Modal
      handleCancel={handleCancel}
      big
      skippable
      className="flex !justify-end w-full h-full !items-end !p-0"
      contentClassName="!rounded-none animate__slideInRight h-screen !px-10"
      {...props}
    >
      <h2 className="w-full font-medium text-[1.75rem] bg-n0  pb-2 ">
        Botella reciclada #{wasteBottleId}
      </h2>

      {loading && (
        <div className="h-full w-full flex items-center justify-center">
          <LoadingSpinner className="mt-8" size="3rem" />
        </div>
      )}

      {!!wasteBottle && (
        <div className="h-full w-full overflow-auto hide-scroll relative pt-2 space-y-6">
          {/* INFORMACIÓN DEL PRODUCTO Y ENVASE */}
          <section className="grid md:grid-cols-2 gap-6 border  border-n2 rounded-xl p-4">
            <h4 className="w-full font-medium  col-span-2 text-neutral-600 ">
              Origen
            </h4>
            {/* Producto */}
            {wasteBottle.product && (
              <article className="space-y-2 ">
                <h3 className="text-lg font-semibold pb-2 border-b">
                  Producto
                </h3>
                <p>
                  <strong>Código de seguimiento:</strong>{' '}
                  {wasteBottle.product.trackingCode}
                </p>
                {wasteBottle.product.ownerName && (
                  <p>
                    <strong>Productor:</strong> {wasteBottle.product.ownerName}
                  </p>
                )}
                <p>
                  <strong>Fecha de producción:</strong>{' '}
                  {new Date(wasteBottle.product.createdAt).toLocaleDateString()}
                </p>
              </article>
            )}

            {/* Envase */}
            {wasteBottle.base && (
              <article className="space-y-2">
                <h3 className="text-lg font-semibold pb-2 border-b">Envase</h3>
                <p>
                  <strong>Peso:</strong> {wasteBottle.base.bottleType.weight}g
                </p>
                <p>
                  <strong>Color:</strong> {wasteBottle.base.bottleType.color}
                </p>
                <p>
                  <strong>Espesor:</strong>{' '}
                  {wasteBottle.base.bottleType.thickness} mm
                </p>
                <p>
                  <strong>Origen:</strong>{' '}
                  {wasteBottle.base.bottleType.originLocation}
                </p>
                {wasteBottle.base.ownerName && (
                  <p>
                    <strong>Productor:</strong> {wasteBottle.base.ownerName}
                  </p>
                )}
              </article>
            )}
            {/* COMPOSICIÓN DEL ENVASE */}
            {wasteBottle.base && (
              <section className="col-span-2">
                <h3 className="text-lg font-semibold pb-2 w-full  border-b">
                  Composición del envase
                </h3>
                {wasteBottle.base.bottleType.composition.length ? (
                  <table className="w-full mt-2 border border-n2 rounded-md">
                    <thead className="bg-n1 text-gray-700">
                      <tr>
                        <th className="px-4 py-2 text-left">Material</th>
                        <th className="px-4 py-2 text-left">Cantidad</th>
                        <th className="px-4 py-2 text-left">Unidad</th>
                      </tr>
                    </thead>
                    <tbody>
                      {wasteBottle.base.bottleType.composition.map(
                        (material: Material, index: number) => (
                          <tr key={index} className="border-t">
                            <td className="px-4 py-2">{material.name}</td>
                            <td className="px-4 py-2">{material.amount}</td>
                            <td className="px-4 py-2">
                              {material.measureUnit}
                            </td>
                          </tr>
                        )
                      )}
                    </tbody>
                  </table>
                ) : (
                  <p className="text-n3 mt-2">
                    No hay información de composición.
                  </p>
                )}
              </section>
            )}
          </section>

          {/* RECICLAJE */}
          {wasteBottle.recycling && (
            <section className="flex flex-col gap-4 border  border-n2 rounded-xl p-4">
              <h4 className="w-full font-medium  col-span-2 text-neutral-600 ">
                Reciclaje
              </h4>
              <div>
                <h3 className=" font-semibold pb-2 mb-4 border-b">
                  Lote de reciclaje
                </h3>
                <div className="grid md:grid-cols-2 gap-4 mt-2">
                  <p>
                    <strong>Material:</strong>{' '}
                    {wasteBottle.recycling.materialType}
                  </p>
                  <p>
                    <strong>Tamaño:</strong> {wasteBottle.recycling.size}
                  </p>
                  <p>
                    <strong>Peso:</strong> {wasteBottle.recycling.weight} kg
                  </p>
                  <p>
                    <strong>Fecha de creación:</strong>{' '}
                    {new Date(
                      wasteBottle.recycling.createdAt
                    ).toLocaleDateString()}
                  </p>
                </div>
              </div>

              {/* Composición reciclaje */}
              <div className="mt-6">
                <h4 className=" font-semibold border-b mb-4 pb-2">
                  Composición
                </h4>
                {wasteBottle.recycling.composition.length > 0 ? (
                  <table className="w-full mt-2 border border-gray-300 rounded-md">
                    <thead className="bg-gray-100 text-gray-700">
                      <tr>
                        <th className="px-4 py-2 text-left">Material</th>
                        <th className="px-4 py-2 text-left">Cantidad</th>
                        <th className="px-4 py-2 text-left">Unidad</th>
                      </tr>
                    </thead>
                    <tbody>
                      {wasteBottle.recycling.composition.map(
                        (material: Material, index: number) => (
                          <tr key={index} className="border-t">
                            <td className="px-4 py-2">{material.name}</td>
                            <td className="px-4 py-2">{material.amount}</td>
                            <td className="px-4 py-2">
                              {material.measureUnit}
                            </td>
                          </tr>
                        )
                      )}
                    </tbody>
                  </table>
                ) : (
                  <p className="text-gray-600 mt-2">
                    No hay información de composición.
                  </p>
                )}
              </div>

              {/* Extra info */}
              {wasteBottle.recycling.extraInfo && (
                <div className="mt-6">
                  <h4 className="text-base font-semibold border-t pt-4 pb-2">
                    Información Adicional
                  </h4>
                  <p className="text-gray-700 mt-2">
                    {wasteBottle.recycling.extraInfo}
                  </p>
                </div>
              )}
            </section>
          )}
        </div>
      )}
    </Modal>
  );
};

export default BottleDetailModal;
