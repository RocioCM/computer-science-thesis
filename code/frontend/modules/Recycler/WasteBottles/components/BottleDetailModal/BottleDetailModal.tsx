import Modal, { ModalProps } from '@/common/components/Modal';
import RecyclerServices from '../../services';
import { toast } from 'react-toastify';
import { Material } from '../../types';
import { useEffect, useState } from 'react';

export interface Props extends ModalProps {
  handleCancel: () => any;
  wasteBottleId: number | null;
}

const BottleDetailModal = ({
  handleCancel,
  wasteBottleId,
  ...props
}: Props) => {
  const [wasteBottle, setWasteBottle] = useState<any | null>(null);

  const fetchWasteBottle = async (wasteBottleId: number) => {
    const { ok, data } = await RecyclerServices.getWasteBottleTracking(
      wasteBottleId
    );
    if (ok) {
      const stagesObject = data.reduce<Record<string, any>>((acc, curr) => {
        acc[curr.stage] = curr.data;
        return acc;
      }, {});
      const { ok, data: dataHistory } = await RecyclerServices.searchBottle(
        stagesObject.pickup?.trackingCode
      );
      if (ok) {
        dataHistory.forEach((item: any) => {
          stagesObject[item.stage] = item.data;
        });
        setWasteBottle(stagesObject);
      } else {
        toast.error(
          'Código no encontrado. Por favor, verifique e intente nuevamente'
        );
      }
    } else {
      toast.error(
        'Código no encontrado. Por favor, verifique e intente nuevamente'
      );
    }
  };

  useEffect(() => {
    if (wasteBottleId) {
      fetchWasteBottle(wasteBottleId);
    }
  }, [wasteBottleId]);

  if (!wasteBottleId) {
    return null;
  }

  return (
    <Modal handleCancel={handleCancel} big skippable {...props}>
      <h2 className="w-full bg-n0">Buscar botellas</h2>

      {!!wasteBottle && (
        <>
          <hr className="w-[95%] border border-n1 my-6 mx-auto" />

          <div className="h-full w-full overflow-auto hide-scroll relative pt-2">
            {/* Información General */}
            <div className="grid grid-cols-2 gap-3xl">
              {wasteBottle.product && (
                <div className=" space-y-2">
                  <h3 className="text-lg font-semibold pb-2">
                    Información del producto
                  </h3>

                  <p>
                    <span className="font-semibold">
                      Código de seguimiento:
                    </span>{' '}
                    {wasteBottle.product.trackingCode}
                  </p>
                  {wasteBottle.product.ownerName && (
                    <p>
                      <span className="font-semibold">Productor:</span>{' '}
                      {wasteBottle.product.ownerName}
                    </p>
                  )}
                  <p>
                    <span className="font-semibold">Fecha de producción:</span>{' '}
                    {new Date(
                      wasteBottle.product.createdAt
                    ).toLocaleDateString()}
                  </p>
                </div>
              )}

              {/* Detalles de la Botella */}
              {wasteBottle.base && (
                <>
                  <div>
                    <h3 className="text-lg font-semibold pb-2">
                      Detalles del envase
                    </h3>
                    <div className="mt-2 space-y-2">
                      <p>
                        <span className="font-semibold">Peso:</span>{' '}
                        {wasteBottle.base.bottleType.weight}g
                      </p>
                      <p>
                        <span className="font-semibold">Color:</span>{' '}
                        {wasteBottle.base.bottleType.color}
                      </p>
                      <p>
                        <span className="font-semibold">Espesor:</span>{' '}
                        {wasteBottle.base.bottleType.thickness} mm
                      </p>
                      <p>
                        <span className="font-semibold">Origen:</span>{' '}
                        {wasteBottle.base.bottleType.originLocation}
                      </p>
                      {wasteBottle.base.ownerName && (
                        <p>
                          <span className="font-semibold">Productor:</span>{' '}
                          {wasteBottle.base.ownerName}
                        </p>
                      )}
                    </div>
                  </div>

                  {/* Composición */}
                  <div className="mt-6 col-span-2">
                    <h3 className="text-lg font-semibold pb-2">Composición</h3>
                    {wasteBottle.base?.bottleType.composition.length ? (
                      <table className="w-full mt-2 border border-n2 rounded-md">
                        <thead>
                          <tr className="bg-n1 text-gray-700">
                            <th className="px-4 py-2 text-left">Material</th>
                            <th className="px-4 py-2 text-left">Cantidad</th>
                            <th className="px-4 py-2 text-left">Unidad</th>
                          </tr>
                        </thead>
                        <tbody>
                          {wasteBottle.base.bottleType.composition.map(
                            (material: Material, index: number) => (
                              <tr key={index} className="border-t">
                                <td className="px-4 py-2">{material.name}</td>
                                <td className="px-4 py-2">{material.amount}</td>
                                <td className="px-4 py-2">
                                  {material.measureUnit}
                                </td>
                              </tr>
                            )
                          )}
                        </tbody>
                      </table>
                    ) : (
                      <p className="text-n3 mt-2">
                        No hay información de composición.
                      </p>
                    )}
                  </div>
                </>
              )}

              {/* Detalles del Reciclaje */}
              {wasteBottle.recycling && (
                <div className="h-full w-full relative pt-2 col-span-2">
                  {/* Información General */}
                  <div className="grid grid-cols-2 gap-m mt-4">
                    <h3 className="text-lg font-semibold border-b py-2 col-span-2">
                      Lote de reciclaje
                    </h3>
                    <p>
                      <span className="font-semibold">Material:</span>{' '}
                      {wasteBottle.recycling.materialType}
                    </p>
                    <p>
                      <span className="font-semibold">Tamaño:</span>{' '}
                      {wasteBottle.recycling.size}
                    </p>
                    <p>
                      <span className="font-semibold">Peso:</span>{' '}
                      {wasteBottle.recycling.weight} kg
                    </p>
                    <p>
                      <span className="font-semibold">Fecha de creación:</span>{' '}
                      {new Date(
                        wasteBottle.recycling.createdAt
                      ).toLocaleDateString()}
                    </p>
                  </div>

                  {/* Composición */}
                  <div className="mt-6">
                    <h3 className="text-lg font-semibold border-t py-2">
                      Composición
                    </h3>
                    {wasteBottle.recycling.composition.length > 0 ? (
                      <table className="w-full mt-2 border border-gray-300 rounded-md">
                        <thead>
                          <tr className="bg-gray-100 text-gray-700">
                            <th className="px-4 py-2 text-left">Material</th>
                            <th className="px-4 py-2 text-left">Cantidad</th>
                            <th className="px-4 py-2 text-left">Unidad</th>
                          </tr>
                        </thead>
                        <tbody>
                          {wasteBottle.recycling.composition.map(
                            (material: Material, index: number) => (
                              <tr key={index} className="border-t">
                                <td className="px-4 py-2">{material.name}</td>
                                <td className="px-4 py-2">{material.amount}</td>
                                <td className="px-4 py-2">
                                  {material.measureUnit}
                                </td>
                              </tr>
                            )
                          )}
                        </tbody>
                      </table>
                    ) : (
                      <p className="text-gray-600 mt-2">
                        No hay información de composición.
                      </p>
                    )}
                  </div>
                  {/* Información Adicional */}
                  {wasteBottle.recycling.extraInfo && (
                    <div className="mt-6">
                      <h3 className="text-lg font-semibold border-t py-2">
                        Información Adicional
                      </h3>
                      <p className="text-gray-700 mt-2">
                        {wasteBottle.recycling.extraInfo}
                      </p>
                    </div>
                  )}
                </div>
              )}
            </div>
          </div>
        </>
      )}
    </Modal>
  );
};

export default BottleDetailModal;
