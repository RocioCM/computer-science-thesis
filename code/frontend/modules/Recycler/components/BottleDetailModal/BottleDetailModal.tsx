import Modal, { ModalProps } from '@/common/components/Modal';
import { toast } from 'react-toastify';
import useForm from '@/common/hooks/useForm';
import { SEARCH_FORM_INPUTS, SEARCH_FORM_STRUCT } from '../../constants';
import Button from '@/common/components/Button';
import { Material } from '../../types';
import RecyclerServices from '../../services';

export interface Props extends ModalProps {
  handleCancel: () => any;
}

const BottleDetailModal = ({ handleCancel, ...props }: Props) => {
  const { form, formBuilder, handleMultipleChange } =
    useForm(SEARCH_FORM_STRUCT);

  const handleSearch = async () => {
    const { ok, data } = await RecyclerServices.searchBottle(form.search);
    if (ok) {
      const stagesObject = data.reduce<Record<string, any>>((acc, curr) => {
        acc[curr.stage] = curr.data;
        return acc;
      }, {});
      handleMultipleChange({ showDetail: true, ...stagesObject }, false);
    } else {
      toast.error(
        'Código no encontrado. Por favor, verifique e intente nuevamente'
      );
    }
  };

  return (
    <Modal handleCancel={handleCancel} big skippable {...props}>
      <h2 className="w-full bg-n0">Buscador de botellas</h2>
      <div className="w-full flex gap-m mt-4">
        {formBuilder(SEARCH_FORM_INPUTS)}
        <Button
          handleClick={handleSearch}
          disabled={!form.search}
          width="auto"
          label="Buscar"
          className="mt-auto"
        />
      </div>
      {form.showDetail && (
        <>
          <hr className="w-[95%] border border-n1 my-6 mx-auto" />

          <div className="h-full w-full overflow-auto hide-scroll relative pt-2">
            {/* Información General */}
            <div className="grid grid-cols-2 gap-3xl">
              {form.product && (
                <div className=" space-y-2">
                  <h3 className="text-lg font-semibold pb-2">
                    Información del producto
                  </h3>

                  <p>
                    <span className="font-semibold">
                      Código de seguimiento:
                    </span>{' '}
                    {form.product.trackingCode}
                  </p>
                  {form.product.ownerName && (
                    <p>
                      <span className="font-semibold">Productor:</span>{' '}
                      {form.product.ownerName}
                    </p>
                  )}
                  <p>
                    <span className="font-semibold">Fecha de producción:</span>{' '}
                    {new Date(form.product.createdAt).toLocaleDateString()}
                  </p>
                </div>
              )}

              {/* Detalles de la Botella */}
              {form.base && (
                <>
                  <div>
                    <h3 className="text-lg font-semibold pb-2">
                      Detalles del envase
                    </h3>
                    <div className="mt-2 space-y-2">
                      <p>
                        <span className="font-semibold">Peso:</span>{' '}
                        {form.base.bottleType.weight}g
                      </p>
                      <p>
                        <span className="font-semibold">Color:</span>{' '}
                        {form.base.bottleType.color}
                      </p>
                      <p>
                        <span className="font-semibold">Espesor:</span>{' '}
                        {form.base.bottleType.thickness} mm
                      </p>
                      <p>
                        <span className="font-semibold">Origen:</span>{' '}
                        {form.base.bottleType.originLocation}
                      </p>
                      {form.base.ownerName && (
                        <p>
                          <span className="font-semibold">Productor:</span>{' '}
                          {form.base.ownerName}
                        </p>
                      )}
                    </div>
                  </div>

                  {/* Composición */}
                  <div className="mt-6 col-span-2">
                    <h3 className="text-lg font-semibold pb-2">Composición</h3>
                    {form.base?.bottleType.composition.length ? (
                      <table className="w-full mt-2 border border-n2 rounded-md">
                        <thead>
                          <tr className="bg-n1 text-gray-700">
                            <th className="px-4 py-2 text-left">Material</th>
                            <th className="px-4 py-2 text-left">Cantidad</th>
                            <th className="px-4 py-2 text-left">Unidad</th>
                          </tr>
                        </thead>
                        <tbody>
                          {form.base.bottleType.composition.map(
                            (material: Material, index: number) => (
                              <tr key={index} className="border-t">
                                <td className="px-4 py-2">{material.name}</td>
                                <td className="px-4 py-2">{material.amount}</td>
                                <td className="px-4 py-2">
                                  {material.measureUnit}
                                </td>
                              </tr>
                            )
                          )}
                        </tbody>
                      </table>
                    ) : (
                      <p className="text-n3 mt-2">
                        No hay información de composición.
                      </p>
                    )}
                  </div>
                </>
              )}
            </div>
          </div>
        </>
      )}
    </Modal>
  );
};

export default BottleDetailModal;
