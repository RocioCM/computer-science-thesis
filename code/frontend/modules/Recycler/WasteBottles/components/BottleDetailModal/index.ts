import BottleDetailModal from './BottleDetailModal';

export type { Props as BottleDetailModalProps } from './BottleDetailModal';
export default BottleDetailModal;
