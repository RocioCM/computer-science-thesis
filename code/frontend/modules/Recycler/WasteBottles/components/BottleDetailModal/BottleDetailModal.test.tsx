import { render } from '@testing-library/react';
import BottleDetailModal from './BottleDetailModal';

describe('Recycler Waste Bottles - BottleDetailModal', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    const modalRoot = document.createElement('div');
    modalRoot.setAttribute('id', 'modal-root');
    document.body.appendChild(modalRoot);
  });

  afterEach(() => {
    const modalRoot = document.getElementById('modal-root');
    if (modalRoot) {
      document.body.removeChild(modalRoot);
    }
  });

  it('Matches snapshot', () => {
    const { container } = render(<BottleDetailModal handleCancel={() => {}} />);
    expect(container).toMatchSnapshot();
  });
});
