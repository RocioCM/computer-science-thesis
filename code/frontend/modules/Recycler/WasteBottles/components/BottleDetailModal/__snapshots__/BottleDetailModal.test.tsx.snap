// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recycler Waste Bottles - BottleDetailModal Matches snapshot 1`] = `<div />`;
